File: 1-Basic technical skills for job interviews/2-Stacks, Queues, and Heaps/queue_implementation.cpp
#include <bits/stdc++.h>

using namespace std;


/// Base class for Queue Operations
template<typename T>
class AbstractQueue {
public:
    virtual void enqueue(T value) = 0;
    virtual void dequeue() = 0;
    virtual T frontElement() = 0;
    virtual int sizeOfQueue() = 0;
    virtual bool isEmpty() = 0;
    virtual void display() = 0;
};

template<typename T>
class QueueOperations : public AbstractQueue<T> {
private:
    vector<T> elements;
    int front;  // Index of the front element
    int rear;   // Index of the rear element

public:
    QueueOperations() : front(-1), rear(-1) {}

    void enqueue(T value) override {
        if (front == -1) {
            front = 0;
            rear = 0;
        } else {
            rear++;
        }
        elements.push_back(value);
    }

    void dequeue() override {
        if (isEmpty()) {
            throw underflow_error("Queue is empty. You can't remove anything.");
        }
        front++;
    }

    T frontElement() override {
        if (isEmpty()) {
            throw underflow_error("Queue is empty.");
        }
        return elements[front];
    }

    int sizeOfQueue() override {
        return rear - front + 1;
    }

    bool isEmpty() override {
        return front > rear;
    }

    void display() override {
        for (int i = front; i <= rear; i++) {
            cout << elements[i] << " ";
        }
        cout << endl;
    }
};

File: 1-Basic technical skills for job interviews/2-Stacks, Queues, and Heaps/queue_implementation.dart
abstract class AbstractQueue<T> {
  void enqueue(T value);
  void dequeue();
  T frontElement();
  int sizeOfQueue();
  bool isEmpty();
  void display();
}

class QueueOperations<T> implements AbstractQueue<T> {
  List<T> elements = [];
  int front = -1; 
  int rear = -1;  

  @override
  void enqueue(T value) {
    if (front == -1) {
      front = 0;
      rear = 0;
    } else {
      rear++;
    }
    elements.add(value);
  }

  @override
  void dequeue() {
    if (isEmpty()) {
      throw StateError("Queue is empty. You can't remove anything.");
    }
    front++;
  }

  @override
  T frontElement() {
    if (isEmpty()) {
      throw StateError("Queue is empty.");
    }
    return elements[front];
  }

  @override
  int sizeOfQueue() {
    return rear - front + 1;
  }

  @override
  bool isEmpty() {
    return front > rear;
  }

  @override
  void display() {
    for (int i = front; i <= rear; i++) {
      print('${elements[i]} ');
    }
    print('');
  }
}

